File: app/src/main/java/com/example/changfeng/taptapword/ConsultWordActivity.java
package com.example.changfeng.taptapword;

import android.app.Activity;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.example.changfeng.taptapword.net.ApiClient;
import com.example.changfeng.taptapword.util.Utils;

import java.util.Calendar;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class ConsultWordActivity extends Activity {

    private static final String TAG = "ConsultWordActivity";

    public static final String TYPE = "type";
    public static final int TYPE_CONSULT = 1;
    public static final int TYPE_COPY = 2;
    private int type = TYPE_CONSULT;

    com.example.changfeng.taptapword.net.result.BaiduResult baiduResult;
    com.example.changfeng.taptapword.net.result.YoudaoResult youdaoResult;
    String query;

    boolean foundResult = false;

    @Bind(R.id.word_consult_layout)
    LinearLayout wordConsultLayout;

    @Bind(R.id.word_edit_text)
    EditText wordEditText;


    @OnClick(R.id.add_new_word_text_view)
    void addNewWord(View view) {
        resetViews();
        String word = wordEditText.getText().toString();
        resultInfoTextView.setVisibility(View.VISIBLE);
        if (Utils.isEnglishWord(word)) {
            resultInfoTextView.setText(getString(R.string.requesting_result_from_internet));
            getResult(word);
        } else {
            resultInfoTextView.setText(getString(R.string.message_not_support_other_language));
        }
    }

    @Bind(R.id.result_info_text_view)
    TextView resultInfoTextView;

    @Bind(R.id.word_result_layout)
    LinearLayout wordResultLayout;
    @Bind(R.id.word_name_text_view)
    TextView wordNameTextView;
    @Bind(R.id.word_phones_text_view)
    TextView wordPhonesTextView;
    @Bind(R.id.word_means_text_view)
    TextView wordMeansTextView;
    @Bind(R.id.word_web_explains_text_view)
    TextView wordWebExplainsTextView;
    @Bind(R.id.youdao_translate_text_view)
    TextView youdaoTranslateTextView;
    @Bind(R.id.baidu_translate_text_view)
    TextView baiduTranslateTextView;
    @Bind(R.id.add_note_text_view)
    TextView addNoteTextView;

    @OnClick(R.id.add_note_text_view)
    void addNote(View view) {
        addNoteTextView.setVisibility(View.GONE);
        noteEditText.setVisibility(View.VISIBLE);
    }

    @Bind(R.id.note_edit_text)
    EditText noteEditText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.activity_consult_word);
        ButterKnife.bind(this);

        type = getIntent().getIntExtra(TYPE, TYPE_CONSULT);
        if (type == TYPE_COPY) {
            wordConsultLayout.setVisibility(View.GONE);
            resultInfoTextView.setVisibility(View.VISIBLE);
            resultInfoTextView.setText(getString(R.string.requesting_result_from_internet));
            String clip = getIntent().getStringExtra(ClipboardService.clipboardText).trim();
            getResult(clip);
        }

        updateViewVisibility();
    }

    @Override
    protected void onDestroy() {
        if (foundResult) {
            saveWord();
        }
        super.onDestroy();
    }

    private void getResult(final String query) {
        foundResult = false;
        this.query = query;
        ApiClient apiClient = new ApiClient();
        final SharedPreferences pref = getSharedPreferences(SharedPref.NAME, MODE_PRIVATE);
        if (pref.getBoolean(SharedPref.BAIDU_TRANSLATE, true)) {
            apiClient.getBaiduResult(query, new Callback<com.example.changfeng.taptapword.net.result.BaiduResult>() {
                @Override
                public void onResponse(Call<com.example.changfeng.taptapword.net.result.BaiduResult> call, Response<com.example.changfeng.taptapword.net.result.BaiduResult> response) {
                    baiduResult = response.body();
                    baiduTranslateTextView.setText(baiduResult.getResult());
                    updateViewVisibility();
                    foundResult = true;
                }

                @Override
                public void onFailure(Call<com.example.changfeng.taptapword.net.result.BaiduResult> call, Throwable t) {
                    resultInfoTextView.setText(getString(R.string.message_cannot_find_word_result));
                }
            });
        }

        apiClient.getYoudaoResult(query, new Callback<com.example.changfeng.taptapword.net.result.YoudaoResult>() {
            @Override
            public void onResponse(Call<com.example.changfeng.taptapword.net.result.YoudaoResult> call, Response<com.example.changfeng.taptapword.net.result.YoudaoResult> response) {
                youdaoResult = response.body();
                resultInfoTextView.setText("");
                if (pref.getBoolean(SharedPref.YOUDAO_DICT, true)) {
                    try {
                        wordPhonesTextView.setText(youdaoResult.getFormatPhones());
                        wordMeansTextView.setText(youdaoResult.getParsedExplains());
                    } catch (Exception e) {

                    }
                }

                if (pref.getBoolean(SharedPref.WEB_EXPLAIN, true)) {
                    try {
                        wordWebExplainsTextView.setText(youdaoResult.getParseWebTranslate());
                    } catch (Exception e) {

                    }
                }

                if (pref.getBoolean(SharedPref.YOUDAO_TRANSLATE, true)) {
                    try {
                        youdaoTranslateTextView.setText(youdaoResult.getTranslateResult());
                    } catch (Exception e) {

                    }
                }
                updateViewVisibility();
                foundResult = true;
            }

            @Override
            public void onFailure(Call<com.example.changfeng.taptapword.net.result.YoudaoResult> call, Throwable t) {
                resultInfoTextView.setText(getString(R.string.message_cannot_find_word_result));
            }
        });


    }

    void showToast(String info) {
        Toast.makeText(getApplicationContext(), info, Toast.LENGTH_SHORT).show();
    }


    private void saveWord() {
        if (query == null || query.isEmpty()) {
            return;
        }
        Word word = new Word();
        word.setLanguage("English");
        word.setName(query);
        word.setEnPhone(youdaoResult.getUkPhonetic());
        word.setAmPhone(youdaoResult.getUsPhonetic());
        try {
            word.setMeans(youdaoResult.getParsedExplains());
        } catch (NullPointerException expected) {
            return;
        }
        word.setWebExplains(youdaoResult.getParseWebTranslate());
        word.setArchived(false);
        word.setNote(noteEditText.getText().toString());

        Calendar c = Calendar.getInstance();
        word.setYear(c.get(Calendar.YEAR));
        word.setMonth(c.get(Calendar.MONTH) + 1);
        word.setDate(c.get(Calendar.DAY_OF_MONTH));
        word.setHour(c.get(Calendar.HOUR_OF_DAY));
        word.setMinute(c.get(Calendar.MINUTE));
        word.setSecond(c.get(Calendar.SECOND));

        WordManger.get(getApplicationContext()).insertWord(word);
    }

    private void updateViewVisibility() {
        boolean wordNameEmpty = wordNameTextView.getText().toString().isEmpty();
        boolean wordPhonesEmpty = wordPhonesTextView.getText().toString().isEmpty();
        boolean wordMeansEmpty = wordMeansTextView.getText().toString().isEmpty();
        boolean wordWebExplainsEmpty = wordWebExplainsTextView.getText().toString().isEmpty();
        boolean youdaoTranslateEmpty = youdaoTranslateTextView.getText().toString().isEmpty();
        boolean baiduTranslateEmpty = baiduTranslateTextView.getText().toString().isEmpty();
        boolean resultInfoEmpty = resultInfoTextView.getText().toString().isEmpty();

        if (wordNameEmpty) {
            wordNameTextView.setVisibility(View.GONE);
        } else {
            wordNameTextView.setVisibility(View.VISIBLE);
        }

        if (wordPhonesEmpty) {
            wordPhonesTextView.setVisibility(View.GONE);
        } else {
            wordPhonesTextView.setVisibility(View.VISIBLE);
        }

        if (wordMeansEmpty) {
            wordMeansTextView.setVisibility(View.GONE);
        } else {
            wordMeansTextView.setVisibility(View.VISIBLE);
        }

        if (wordWebExplainsEmpty) {
            wordWebExplainsTextView.setVisibility(View.GONE);
        } else {
            wordWebExplainsTextView.setVisibility(View.VISIBLE);
        }

        if (youdaoTranslateEmpty) {
            youdaoTranslateTextView.setVisibility(View.GONE);
        } else {
            youdaoTranslateTextView.setVisibility(View.VISIBLE);
        }

        if (baiduTranslateEmpty) {
            baiduTranslateTextView.setVisibility(View.GONE);
        } else {
            baiduTranslateTextView.setVisibility(View.VISIBLE);
        }

        if (wordPhonesEmpty && wordMeansEmpty && wordWebExplainsEmpty && youdaoTranslateEmpty && baiduTranslateEmpty) {
            addNoteTextView.setVisibility(View.GONE);
        } else {
            addNoteTextView.setVisibility(View.VISIBLE);
        }

        if (resultInfoEmpty) {
            resultInfoTextView.setVisibility(View.GONE);
        } else {
            resultInfoTextView.setVisibility(View.VISIBLE);
        }


    }

    private void resetViews() {
        wordNameTextView.setText("");
        wordPhonesTextView.setText("");
        wordMeansTextView.setText("");
        wordWebExplainsTextView.setText("");
        youdaoTranslateTextView.setText("");
        baiduTranslateTextView.setText("");
        noteEditText.setText("");
        resultInfoTextView.setText("");
        updateViewVisibility();
    }


}

File: app/src/main/java/com/example/changfeng/taptapword/ConsultWordActivity.kt
package com.example.changfeng.taptapword

import android.app.Activity
import android.app.PendingIntent
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.Window
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.app.NotificationCompat
import com.example.changfeng.taptapword.net.ApiClient
import com.example.changfeng.taptapword.net.result.BaiduResult
import com.example.changfeng.taptapword.net.result.YoudaoResult
import com.example.changfeng.taptapword.util.LogUtils
import com.example.changfeng.taptapword.util.isEnglishWord
import com.umeng.analytics.MobclickAgent
import org.jetbrains.anko.defaultSharedPreferences
import org.jetbrains.anko.find
import org.jetbrains.anko.notificationManager
import org.jetbrains.anko.sdk27.coroutines.onClick
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*


class ConsultWordActivity : Activity() {
    private var type = TYPE_CONSULT

    var baiduResult: BaiduResult? = null
    var youdaoResult: YoudaoResult? = null
    var query: String? = null

    var foundResult = false


    val wordConsultLayout: LinearLayout get() = find(R.id.word_consult_layout)
    val wordEditText: TextView get() = find(R.id.word_edit_text)
    val addNewTextView: TextView get() = find(R.id.add_new_word_text_view)
    val resultInfoTextView: TextView get() = find(R.id.result_info_text_view)
    val wordResultLayout: LinearLayout get() = find(R.id.word_result_layout)
    val wordNameTextView: TextView get() = find(R.id.word_name_text_view)
    val wordPhonesTextView: TextView get() = find(R.id.word_phones_text_view)
    val wordMeansTextView: TextView get() = find(R.id.word_means_text_view)
    val wordWebExplainsTextView: TextView get() = find(R.id.word_web_explains_text_view)
    val youdaoTranslateTextView: TextView get() = find(R.id.youdao_translate_text_view)
    val baiduTranslateTextView: TextView get() = find(R.id.baidu_translate_text_view)
    val addNoteTextView: TextView get() = find(R.id.add_note_text_view)
    val noteEditText: TextView get() = find(R.id.note_edit_text)


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        LogUtils.LOGI(TAG, "onCreate()")
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.activity_consult_word)
        addNewTextView.onClick {
            resetViews()
            val word = wordEditText.text.toString()
            resultInfoTextView.visibility = View.VISIBLE
            if (word.isEnglishWord()) {
                resultInfoTextView.text = getString(R.string.requesting_result_from_internet)
                getResult(word)
            } else {
                resultInfoTextView.text = getString(R.string.message_not_support_other_language)
            }
        }

        addNoteTextView.onClick {
            addNoteTextView.visibility = View.GONE
            noteEditText.visibility = View.VISIBLE
        }

        type = intent.getIntExtra(TYPE, TYPE_CONSULT)


        if (type == TYPE_COPY) {
            wordConsultLayout.visibility = View.GONE
            resultInfoTextView.visibility = View.VISIBLE
            resultInfoTextView.text = getString(R.string.requesting_result_from_internet)
            val clip = intent.getStringExtra(ClipboardService.clipboardText).trim()
            getResult(clip)
        }

        updateViewVisibility()
    }

    override fun onResume() {
        super.onResume()
        MobclickAgent.onPageStart("网络查询单词页面")
        MobclickAgent.onResume(this)
    }

    override fun onPause() {
        super.onPause()
        if (defaultSharedPreferences.getBoolean(SharedPref.NOTIFICATION_NEW_WORD, true) && !wordNameTextView.text.isNullOrEmpty()) {
            val notificationBuilder = NotificationCompat.Builder(this).setSmallIcon(R.drawable.image_ninja).setContentTitle(getString(R.string.app_name)).setContentText("新单词 ${wordNameTextView.text.toString()}")
            val i = Intent(this, MainActivity::class.java).putExtra(MainActivity.NINJA_NEW_WORD, true)
            val pi = PendingIntent.getActivity(this, 0, i, PendingIntent.FLAG_CANCEL_CURRENT)
            notificationBuilder.setContentIntent(pi)
            notificationManager.notify(1, notificationBuilder.build())
        }

        MobclickAgent.onPageEnd("网络查询单词页面")
        MobclickAgent.onPause(this)
    }


    override fun onDestroy() {
        if (foundResult) {
            saveWord()
        }

        super.onDestroy()
    }

    private fun getResult(query: String) {
        LogUtils.LOGI(TAG, "getResult() $query")
        foundResult = false
        this.query = query
        val apiClient = ApiClient()
        val pref = defaultSharedPreferences
        if (pref.getBoolean(SharedPref.BAIDU_TRANSLATE, true)) {
            apiClient.getBaiduResult(query, object : Callback<BaiduResult> {
                override fun onResponse(call: Call<BaiduResult>, response: Response<BaiduResult>) {
                    baiduResult = response.body()
                    baiduTranslateTextView.text = baiduResult?.result
                    updateViewVisibility()
                    foundResult = true
                }

                override fun onFailure(call: Call<BaiduResult>, t: Throwable) {
                    t.printStackTrace()
                    resultInfoTextView.text = getString(R.string.message_cannot_find_word_result)
                }
            })
        }

        apiClient.getYoudaoResult(query, object : Callback<YoudaoResult> {
            override fun onResponse(call: Call<YoudaoResult>, response: Response<YoudaoResult>) {

                youdaoResult = response.body()
                wordNameTextView.text = query
                resultInfoTextView.text = ""
                if (pref.getBoolean(SharedPref.YOUDAO_DICT, true)) {
                    try {
                        LogUtils.LOGI(TAG, "onResponse() $query  phone:${youdaoResult?.formatPhones}")
                        wordPhonesTextView.text = youdaoResult?.formatPhones
                        wordMeansTextView.text = youdaoResult?.parsedExplains
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }

                if (pref.getBoolean(SharedPref.WEB_EXPLAIN, true)) {
                    try {
                        wordWebExplainsTextView.text = youdaoResult?.parseWebTranslate
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }

                if (pref.getBoolean(SharedPref.YOUDAO_TRANSLATE, true)) {
                    try {
                        youdaoTranslateTextView.text = youdaoResult?.translateResult
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }
                updateViewVisibility()
                foundResult = true
            }

            override fun onFailure(call: Call<YoudaoResult>, t: Throwable) {
                t.printStackTrace()
                resultInfoTextView.text = getString(R.string.message_cannot_find_word_result)
            }
        })


    }

    private fun saveWord() {
        if (query == null || query!!.isEmpty()) {
            return
        }
        val word = Word()
        word.language = "English"
        word.name = query
        word.enPhone = youdaoResult?.ukPhonetic
        word.amPhone = youdaoResult?.usPhonetic
        try {
            word.means = youdaoResult?.parsedExplains
        } catch (expected: NullPointerException) {
            return
        }

        word.webExplains = youdaoResult?.parseWebTranslate
        word.isArchived = false
        word.note = noteEditText.text.toString()

        val c = Calendar.getInstance()
        word.year = c.get(Calendar.YEAR)
        word.month = c.get(Calendar.MONTH) + 1
        word.date = c.get(Calendar.DAY_OF_MONTH)
        word.hour = c.get(Calendar.HOUR_OF_DAY)
        word.minute = c.get(Calendar.MINUTE)
        word.second = c.get(Calendar.SECOND)

        WordManger.get(applicationContext).insertWord(word)
    }

    private fun updateViewVisibility() {
        wordNameTextView.visibility = if (wordNameTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        wordPhonesTextView.visibility = if (wordPhonesTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        wordMeansTextView.visibility = if (wordMeansTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        wordWebExplainsTextView.visibility = if (wordWebExplainsTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        youdaoTranslateTextView.visibility = if (youdaoTranslateTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        baiduTranslateTextView.visibility = if (baiduTranslateTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        addNewTextView.visibility = if (addNoteTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        resultInfoTextView.visibility = if (resultInfoTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
        addNoteTextView.visibility = if (wordPhonesTextView.text!!.isEmpty() && wordMeansTextView.text!!.isEmpty() && youdaoTranslateTextView.text!!.isEmpty() && baiduTranslateTextView.text!!.isEmpty()) View.GONE else View.VISIBLE
    }

    private fun resetViews() {
        wordNameTextView.text = ""
        wordPhonesTextView.text = ""
        wordMeansTextView.text = ""
        wordWebExplainsTextView.text = ""
        youdaoTranslateTextView.text = ""
        baiduTranslateTextView.text = ""
        noteEditText.text = ""
        resultInfoTextView.text = ""
        updateViewVisibility()
    }

    companion object {

        private const val TAG = "ConsultWordActivity"

        const val TYPE = "type"
        const val TYPE_CONSULT = 1
        const val TYPE_COPY = 2
    }


}
